### ISSUE TITLE
cri-rm did not apply cpuset.mems to the process

### ISSUE THREAD

**Issue by wang-h-ng** [intel/cri-resource-manager#878]
CRI-RM uses the topology-aware policy. The log shows that redis is fixed to memory 1, but I use numastat - P to find that it is also used on memory 0:

### logs:
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [      cache     ] saving cache to file '/var/lib/cri-resmgr/cache'...
**Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ] * applying grant <grant for redis-75655f478b-h8qhj:redis from socket #1: cputype: normal>
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ]   => pinning to (shared) cpuset 18-35,54-71
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ]   => pinning to memory 1**
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ] redis-75655f478b-h8qhj:redis: setting demotion preferences...
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ] * updating shared allocations affected by <grant for redis-75655f478b-h8qhj:redis from socket #1: cputype: normal>
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ]   => updating <grant for calico-node-rb4qx:calico-node from socket #1: cputype: normal> with shared CPUs of socket #1: 18-35,54-71...
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ]   => updating <grant for calico-typha-ddd545dd6-fmwhn:calico-typha from socket #0: cputype: normal> with shared CPUs of socket #0: 1-17,36-53...
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ]   => updating <grant for csi-node-driver-gmknn:csi-node-driver-registrar from socket #0: cputype: normal> with shared CPUs of socket #0: 1-17,36-53...
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ]   => <grant for kube-proxy-5vqht:kube-proxy from root: cputype: reserved> not affected (only reserved CPUs)...
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ]   => updating <grant for csi-node-driver-gmknn:calico-csi from socket #1: cputype: normal> with shared CPUs of socket #1: 18-35,54-71...
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ]   => updating <grant for tigera-operator-7ff575f7f7-nlw67:tigera-operator from socket #0: cputype: normal> with shared CPUs of socket #0: 1-17,36-53...
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ] <post-alloc><virtual root>
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ] <post-alloc>  - <root capacity: CPU: reserved:0 (1000m), sharable:1-71 (71000m), MemLimit: DRAM 187.62G>
Sep 02 07:16:09 stx002 cri-resmgr[279067]: D: [     policy     ] <post-alloc>  - <root allocatable: CPU: reserved:0 (allocatable: 1000m), sharable:1-71 (allocatable:71000m), MemLimit: DRAM 187.62G>

### nestat command display:
root@stx002:/dev/shm# numastat -p 674154
Per-node process memory usage (in MBs) for PID 674154 (redis-server)
                           Node 0               Node 1           Total
                  --------------- --------------- ---------------
Huge                         0.00               0.00            0.00
Heap                         0.00               0.19            0.19
Stack                         0.00               0.02            0.02
Private                      **5.07**               4.02             9.09
----------------  --------------- --------------- ---------------
Total                          5.07              4.23             9.30

### cri-resmgr version information:
  - version: 0.7.0
  - build:   647cd55

kubernetes version:
v1.24.0
os:
Ubuntu 20.04.3 LTS
kenel:
5.4.0-124-generic
containerd version:
1.6.8

**Comment by kad** [comment]
What shows `grep allowed_list /proc/674154/status` ? also, it is good to look at `/proc/674154/numa_maps`. Often, memory marked as private belongs to memory mapped files (binary itself, all dependant shared libraries). Those files are loaded only once by kernel, for many instances of the same workloads. Those are not following restrictions of memset.

**Comment by wang-h-ng** [comment]
Thank you. I think it's the above reasons that cause my test data to be poor. It doesn't seem to be the case.

I am using memtier_ Benchmark tests redis, but the test results are not very good. The details are as follows:

### Log of deploying redis:
Sep 02 08:28:41 jfz1r03h21 cri-resmgr[1606278]: D: [     cache     ] saving cache to file '/var/lib/cri-resmgr/cache'...
Sep 02 08:28:41 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ] * applying grant <grant for redis-598654c75-xh4km:redis from socket #1: cputype: normal>
Sep 02 08:28:41 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ]   => pinning to (shared) cpuset 24-47,72-95
Sep 02 08:28:41 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ]   => pinning to memory 1
Sep 02 08:28:41 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ] redis-598654c75-xh4km:redis: setting demotion preferences...
### Deploy memtier_ Benchmark's log(Affinity is added to the annotations of pod):
Sep 02 08:39:34 jfz1r03h21 cri-resmgr[1606278]: D: [      cache     ] saving cache to file '/var/lib/cri-resmgr/cache'...
Sep 02 08:39:34 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ] * applying grant <grant for memtier-benchmark-k2h9n:memtier-benchmark from socket #1: cputype: normal>
Sep 02 08:39:34 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ]   => pinning to (shared) cpuset 24-47,72-95
Sep 02 08:39:34 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ]   => pinning to memory 1

### test result:
ALL STATS
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets         4055.14          ---          ---         4.48522         4.47900         8.95900        15.93500       312.36
Gets        40542.76        20.00     40522.77         4.48182         4.47900         8.95900        15.80700      1579.90
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals      44597.90        20.00     40522.77         4.48213         4.47900         8.95900        15.80700      1892.26

### Deploy memtier_ Benchmark's log:(Anti affinity is added to the annotations of pod)
Sep 02 08:44:36 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ] * applying grant <grant for memtier-benchmark-mz87p:memtier-benchmark from socket #0: cputype: normal>
Sep 02 08:44:36 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ]   => pinning to (shared) cpuset 1-23,48-71
Sep 02 08:44:36 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ]   => pinning to memory 0
Sep 02 08:44:36 jfz1r03h21 cri-resmgr[1606278]: D: [     policy     ] memtier-benchmark-mz87p:memtier-benchmark: setting demotion preferences...

### test result:
ALL STATS
============================================================================================================================
Type         Ops/sec     Hits/sec   Misses/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
----------------------------------------------------------------------------------------------------------------------------
Sets         4355.13          ---          ---         4.17541         4.25500         8.38300        16.76700       335.48
Gets        43542.91        25.00     43517.91         4.17326         4.25500         8.38300        16.76700      1696.94
Waits           0.00          ---          ---             ---             ---             ---             ---          ---
Totals      47898.04        25.00     43517.91         4.17345         4.25500         8.38300        16.76700      2032.42

lscpu command output:
NUMA node0 CPU(s):               0-23,48-71
NUMA node1 CPU(s):               24-47,72-95

**Comment by wang-h-ng** [comment]
The above test results are a bit confusing. Let me add it again:
### test result:(Affinity is added to the annotations of pod)
![image](https://user-images.githubusercontent.com/84434914/188104474-0b77ad3c-8b95-4a65-86f9-1685abd6eb74.png)
### test result:(Anti affinity is added to the annotations of pod)
![image](https://user-images.githubusercontent.com/84434914/188104637-68e8fb2d-fe96-4772-8245-3a95aaa1a4e2.png)

**Comment by askervin** [comment]
Hi @wang-h-ng,

1. Would you share which memory nodes are allowed for memtier_benchmark and redis-server processes? This makes sure that CPU and memory pinning by the topology aware policy gets correctly passed down to kernel. Example, taken when the test is running:

```
root@n4c16-debian-sid-crirm-containerd:/home/debian# grep _allowed_list /proc/$(pidof memtier_benchmark)/status
Cpus_allowed_list:      4-7
Mems_allowed_list:      1
root@n4c16-debian-sid-crirm-containerd:/home/debian# grep _allowed_list /proc/$(pidof redis-server)/status
Cpus_allowed_list:      4-7
Mems_allowed_list:      1
```

2. If that looks expected, then let's see NUMA maps of these processes. It will reveal which parts of the memory that the processes use are located on unexpected NUMA nodes:
```
root@n4c16-debian-sid-crirm-containerd:/home/debian# cat /proc/$(pidof redis-server)/numa_maps
55eb1344d000 default file=/usr/local/bin/redis-server mapped=107 N3=107 kernelpagesize_kB=4
55eb134b8000 default file=/usr/local/bin/redis-server mapped=369 N3=369 kernelpagesize_kB=4
55eb13639000 default file=/usr/local/bin/redis-server mapped=71 N3=71 kernelpagesize_kB=4
55eb136bc000 default file=/usr/local/bin/redis-server anon=5 dirty=5 mapped=7 active=2 N1=5 N3=2 kernelpagesize_kB=4
55eb136c3000 default file=/usr/local/bin/redis-server anon=90 dirty=90 active=0 N1=90 kernelpagesize_kB=4
55eb1371d000 default anon=42 dirty=42 active=0 N1=42 kernelpagesize_kB=4
55eb13b27000 default heap anon=48 dirty=48 active=0 N1=48 kernelpagesize_kB=4
7fe3312fc000 default anon=53 dirty=53 active=0 N1=53 kernelpagesize_kB=4
...
```

**Comment by wang-h-ng** [comment]
hi @askervin 
    The following is the information of redis and memtier-benchmark (affinity is set):
### redis:
root@jfz1r03h21:~# grep _allowed_list /proc/3258371/status
Cpus_allowed_list:      24-47,72-95
Mems_allowed_list:      1

root@jfz1r03h21:~# cat /proc/3258371/numa_maps
55baec321000 default file=/usr/local/bin/redis-server mapped=107 N1=107 kernelpagesize_kB=4
55baec38c000 default file=/usr/local/bin/redis-server mapped=353 active=337 N1=353 kernelpagesize_kB=4
55baec50d000 default file=/usr/local/bin/redis-server mapped=67 N1=67 kernelpagesize_kB=4
55baec590000 default file=/usr/local/bin/redis-server anon=5 dirty=5 mapped=7 N1=7 kernelpagesize_kB=4
55baec597000 default file=/usr/local/bin/redis-server anon=90 dirty=90 N1=90 kernelpagesize_kB=4
55baec5f1000 default anon=42 dirty=42 N1=42 kernelpagesize_kB=4
55baecc63000 default heap anon=48 dirty=48 N1=48 kernelpagesize_kB=4
7fd0c5aff000 default anon=8678 dirty=8678 N1=8678 kernelpagesize_kB=4
7fd0c8bff000 default
7fd0c8c00000 default anon=4 dirty=4 N1=4 kernelpagesize_kB=4
7fd0c9400000 default anon=338 dirty=338 N1=338 kernelpagesize_kB=4
7fd0ca3fd000 default
7fd0ca3fe000 default anon=3 dirty=3 N1=3 kernelpagesize_kB=4
7fd0cabfe000 default
7fd0cabff000 default anon=3 dirty=3 N1=3 kernelpagesize_kB=4



### memtier-benchmark:
root@jfz1r03h21:~# grep _allowed_list /proc/3279560/status
Cpus_allowed_list:      24-47,72-95
Mems_allowed_list:      1

root@jfz1r03h21:~# cat /proc/3279560/numa_maps
55cfebb23000 default file=/usr/local/bin/memtier_benchmark mapped=61 N1=61 kernelpagesize_kB=4
55cfebd63000 default file=/usr/local/bin/memtier_benchmark anon=1 dirty=1 N1=1 kernelpagesize_kB=4
55cfebd64000 default file=/usr/local/bin/memtier_benchmark anon=1 dirty=1 N1=1 kernelpagesize_kB=4
55cfebe47000 default heap anon=4101 dirty=4101 N1=4101 kernelpagesize_kB=4
7ffaf8000000 default anon=570 dirty=570 N1=570 kernelpagesize_kB=4
7ffaf823a000 default
7ffafc000000 default anon=569 dirty=569 N1=569 kernelpagesize_kB=4
7ffafc239000 default
7ffb00000000 default anon=571 dirty=571 N1=571 kernelpagesize_kB=4

**Comment by wang-h-ng** [comment]
Hi, my kubelet and containerd use systemd instead of cgroupfs. I don't know if it is related to this

**Comment by askervin** [comment]
@wang-h-ng, `_allowed_list` lines in `/proc/PID/status` look good. Both redis and memtier-benchmark have been pinned to CPUs and memories on the same NUMA node, which is expected in case of the affinity. This means that CPU and memory pinning request flows from CRI-RM through containerd down to cgroups (to both `cpuset.cpus` and `cpuset.mems`) without problems in your system.

`numa_maps` look good, too: all `dirty` and anonymous memory pages of both processes are located on NUMA node 1 (`N1`). And in this case you have been lucky: even `mapped` pages happen to be mapped on the same node. Linux kernel memory policy does not apply to `mapped` (named) pages. These pages can be shared among many processes, and the processes that share them can be pinned to different NUMA nodes. Therefore it may happen that these pages are located on a different NUMA node than where process memory is pinned to. This is the case in my `numa_maps` example in [the earlier comment](https://github.com/intel/cri-resource-manager/issues/878#issuecomment-1236658707).

The original finding in this issue is: `numastat -p PID` shows some of the process memory being on a wrong NUMA node. This is explained by the mapped memory. For example, when investigating my `numa_maps` example deeper:

`numastat -p PID` shows that some of my redis-server process memory is on NUMA node 3, although CRI-RM has pinned the process to NUMA node 1:
```
root@n4c16-debian-sid-crirm-containerd:~# numastat -p  $(pidof redis-server)                                                                                        
Per-node process memory usage (in MBs) for PID 125040 (redis-server)
                           Node 0          Node 1          Node 2
                  --------------- --------------- ---------------
Huge                         0.00            0.00            0.00
Heap                         0.00            0.19            0.00
Stack                        0.00            0.02            0.00
Private                      0.00           10.58            0.00
----------------  --------------- --------------- ---------------
Total                        0.00           10.79            0.00

                           Node 3           Total
                  --------------- ---------------
Huge                         0.00            0.00
Heap                         0.00            0.19
Stack                        0.00            0.02
Private                      2.11           12.69
----------------  --------------- ---------------
Total                        2.11           12.90
```

Let's investigate what memory is in NUMA node 3:
```
root@n4c16-debian-sid-crirm-containerd:~# cat /proc/$(pidof redis-server)/numa_maps | grep N3                                                                       5608c1fb5000 default file=/usr/local/bin/redis-server mapped=107 N3=107 kernelpagesize_kB=4
5608c2020000 default file=/usr/local/bin/redis-server mapped=369 N3=369 kernelpagesize_kB=4
5608c21a1000 default file=/usr/local/bin/redis-server mapped=63 N3=63 kernelpagesize_kB=4
5608c2224000 default file=/usr/local/bin/redis-server anon=5 dirty=5 mapped=7 active=2 N1=5 N3=2 kernelpagesize_kB=4
```

So it seems that they are all memory mapped files, 4 kB memory per page. Just to be sure, let's sum up the amount of this memory to see if this explains all the memory that is reported by `numastat -p PID` on NUMA node 3:
```
root@n4c16-debian-sid-crirm-containerd:~# cat /proc/$(pidof redis-server)/numa_maps | grep N3 | awk 'BEGIN{RS=" ";FS="="}/N3/{print; n+=$2}END{print n" pages, "n*4 /1024" MB"}'
N3=107
N3=369
N3=63
N3=2
541 pages, 2.11328 MB
```

That is, the total MB matches exactly what `numamaps` reports.

Do you think we can close this issue? Or maybe reword it, if the current title does not match the problem you are seeing?
